File: 51.n-皇后.cpp
/*
 * @lc app=leetcode.cn id=51 lang=cpp
 *
 * [51] N 皇后
 */

// @lc code=start
class Solution {
public:
    vector<string> board;
    vector<vector<string>> res;
    vector<bool> arr1;
    vector<bool> arr2;
    vector<bool> cols;
    vector<vector<string>> solveNQueens(int n) {
        arr1.resize(2 * n - 1, false); // r + c
        arr2.resize(2 * n - 1, false); // r - c 's index should consider the edge case is the (n - 1)
        cols.resize(n, false); // 
        board.resize(n, string(n, '.'));
        backtrack(n, 0);
        return res;
    }

    void backtrack(int n, int row) {
        if(row == n) {
            res.push_back(board);
            return;
        }

        for(int c = 0; c < n; c++) {
            if(!arr1[row + c] && !arr2[row - c + n - 1] && !cols[c] && board[row][c] == '.') {
                arr1[row + c] = arr2[row - c + n - 1] = cols[c] = true;
                board[row][c] = 'Q';
                backtrack(n, row + 1);
                board[row][c] = '.';
                arr1[row + c] = arr2[row - c + n - 1] = cols[c] = false;
            }
        }

    }
};
// @lc code=end

File: 51.n-皇后.py
#
# @lc app=leetcode.cn id=51 lang=python3
#
# [51] N 皇后
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []
        # 主对角线和副对角线
        arr1 = [False] * (2 * n - 1) 
        arr2 = [False] * (2 * n - 1) 
        cols = [False] * (n) 
        board = [['.' for _ in range(n)] for _ in range(n)]
        def backtrack(r: int):
            if r == n:
                res.append([''.join(board[i]) for i in range(len(board))])
                return
            
            for col in range(n):
                if arr1[col + r] or arr2[r - col + n - 1] or cols[col]:
                    continue
                board[r][col] = 'Q'
                arr1[r + col] = True
                arr2[r - col + n - 1] = True
                cols[col] = True
                backtrack(r + 1)
                board[r][col] = '.'
                cols[col] = False
                arr1[r + col] = False
                arr2[r - col + n - 1] = False
        backtrack(0)
        return res

                    
     
            




            
                    




        
# @lc code=end

